### PR TITLE
CSHARP-3821: Sync LB test sdam-error-handling.json with the latest spec.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs [hunk 1]
```diff
@@ -241,14 +241,27 @@ public IConnectionHandle AcquireConnection(CancellationToken cancellationToken)
...
                         {
-                            pooledConnection = connectionCreator.CreateOpenedOrReuse(cancellationToken);
+                            try
+                            {
+                                pooledConnection = connectionCreator.CreateOpenedOrReuse(cancellationToken);
+                            }
+                            catch (Exception ex)
+                            {
+                                // we need to apply SDAM rules before connection closing
+                                _pool._connectionExceptionHandler.HandleExceptionOnOpen(ex);
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs:251]: now we need to emit `poolClearedEvent` before `connectionClosedEvent`. We can discuss it

Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs:251]: We can dispose `ConnectionCreator` in `Finally` instead.

File: tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/tests/pool-create-min-size-error.json [hunk 2]
```diff
@@ -49,15 +49,15 @@
       "type": "ConnectionCreated",
       "address": 42
     },
+    {
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/tests/pool-create-min-size-error.json:52]: this change is described here https://jira.mongodb.org/browse/DRIVERS-1785?focusedCommentId=4066485&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4066485, without it, the mentioned test fails

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs [hunk 3]
```diff
@@ -56,6 +56,32 @@ private Exception CreateTimeoutException(Stopwatch stopwatch, string message)
             return new TimeoutException(message);
         }
 
+        private T WithSdamErrorHandling<T>(Func<T> func)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.Helpers.cs:59]: I think it's easiest and the most good looking way to do this logic, one of reasons for this, now we don't need to investigate where SDAM handling happens and where it's not

**Review by BorisDog — APPROVED** [review]
LGTM

**Pull request merged**